data_steward/analytics/cdr_ops/incremental_load_qc.py

```diff
@@ -0,0 +1,503 @@
+        "There are <b>{count}</b> tables that might have "
+        "AIAN participants' records. <br>Look at these tables and investigate why AIAN "
+        "records are there: <b>{tables}</b><br><br>")
+
+    render_message(df,
+                   success_null_check,
+                   failure_null_check,
+                   failure_msg_args={
+                       'count': len(df),
+                       'tables': ', '.join(df.table_name)
+                   })
+# -
+
+# ## QC 2. Confirm there are no duplicate `OBSERVATION_ID`s
+# Any observation records from `incremental_dataset` got new `OBSERVATION_ID`s
+# assigned in `new_dataset` to avoid ID duplicates. <br>We must confirm all the
+# `OBSERVATION_ID`s in `new_dataset` are unique. <br>The same goes to
+# `_mapping_observation` and `obsesrvation_ext`.
+
+# +
+if is_combined_release_dataset(new_dataset):
+    obs_tables = [OBSERVATION, ext_table_for(OBSERVATION)]
+elif is_combined_release_dataset(new_dataset):
+    obs_tables = [
+        OBSERVATION,
+        mapping_table_for(OBSERVATION),
+        ext_table_for(OBSERVATION)
+    ]
+else:
+    obs_tables = [OBSERVATION]
+
+query = JINJA_ENV.from_string('''
+    SELECT
+        \'{{table}}\' as table_name,
+        observation_id,
+        COUNT(*) AS duplicate_row_count
+    FROM `{{project}}.{{new_dataset}}.{{table}}`
+    GROUP BY table_name, observation_id
+    HAVING COUNT(*) > 1
+''')
+
+queries = [
+    query.render(project=project_id, new_dataset=new_dataset, table=table)
+    for table in obs_tables
+]
+
+union_all_query = '\nUNION ALL\n'.join(queries)
+df = execute(client, union_all_query)
+
+success_null_check = f"No OBSERVATION_ID duplicates are found in {new_dataset}.<br><br>"
+failure_null_check = (
+    "There are total <b>{count}</b> OBSERVATION_IDs that might have"
+    "duplicates. <br>Look at these tables and investigate why there are duplicates: <b>{tables}</b><br><br>"
+)
+
+render_message(df,
+               success_null_check,
+               failure_null_check,
+               failure_msg_args={
+                   'count': len(df),
+                   'tables': ', '.join(df.table_name)
+               })
+# -
+
+# ## QC 3. Confirm mapping/ext tables are consistent
+# This hotfix runs several delete and insert statements to OBSERVATION, SURVEY_CONDUCT, PERSON
+# and their mapping/ext tables. <br>We must confirm that mapping/ext tables are consistent with
+# their correspondants after the hotfix.
+
+# +
+if is_combined_release_dataset(new_dataset):
+    map_ext_tuples = [
+        (OBSERVATION, mapping_table_for(OBSERVATION)),
+        (OBSERVATION, ext_table_for(OBSERVATION)),
+        (SURVEY_CONDUCT, mapping_table_for(SURVEY_CONDUCT)),
+        (SURVEY_CONDUCT, ext_table_for(SURVEY_CONDUCT)),
+    ]
+elif is_deid_dataset(new_dataset):
+    map_ext_tuples = [
+        (OBSERVATION, ext_table_for(OBSERVATION)),
+        (SURVEY_CONDUCT, ext_table_for(SURVEY_CONDUCT)),
+        (PERSON, ext_table_for(PERSON)),
```

Comment:
Duplicate comment from `remediate_basics.py`:

Hi, @hiro-mishima - overall this is a good logic addition.  For `person_ext` specifically, it will not exist until `deid_base` or `deid_clean`, and won't exist in `deid`.  Perhaps we should have another function added to `retract_utils.py` like `is_deid_release` that would indicate we are in `deid_base` or `deid_clean`.  Then this new function should be added as a condition to the QC 3 logic check.

Reply:
Thanks for pointing it out! I just learned that `person_ext` exist only in `deid_base` and `deid_clean`, not in `deid`. I updated the logic here: https://github.com/all-of-us/curation/pull/1489/commits/3318e721980ff82c5949fb517012edb6dfa5c98d#diff-69f3f7c70cd2449d5408e740fc8181f307d17b867c49f13fccaabb76825847b5R308